extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -100,17 +103,68 @@ export class AmpAdExit extends AMP.BaseElement {
       'CLICK_Y': true,
     };
     if (target.vars) {
-      for (const customVar in target.vars) {
-        if (customVar[0] == '_') {
-          vars[customVar] = () =>
-              args[customVar] || target.vars[customVar].defaultValue;
-          whitelist[customVar] = true;
+      const all3pResponses = this.getAmpDoc().win[AMP_ANALYTICS_3P_RESPONSES] ||
+          {};
+
+      for (const customVarName in target.vars) {
+        if (customVarName[0] == '_') {
+          const customVar =
+              /** @type {./config.Variable} */ (target.vars[customVarName]);
+          if (customVar) {
+            /*
+              Example:
+              The amp-ad-exit target has a variable representing the
+               priority of something, which is defined as follows:
+               "vars": {
+                 "_pty": {
+                   "defaultValue": "unknown",
+                   "vendorAnalyticsSource": "vendorXYZ",
+                   "vendorAnalyticsResponseKey": "priority"
+                 },
+                 ...
+               }
+               The cross-domain iframe of vendorXYZ has sent the
+               following response for the creative:
+                 { priority: medium, category: W }
+               This is just example data. The keys/values in that object can
+               be any strings.
+               The code below will create substitutionFunctions['_pty'],
+               which in this example will return "medium".
+             */
+            substitutionFunctions[customVarName] = () => {
+              if (customVar.hasOwnProperty('vendorAnalyticsSource') &&
+                  customVar.hasOwnProperty('vendorAnalyticsResponseKey')) {
+                // It's a 3p analytics variable
+                const vendor =
+                    /** @type {string} */ (customVar.vendorAnalyticsSource);
+                if (all3pResponses[vendor]) {
+                  /* The vendor (in the example above, "vendorXYZ") has
+                     responded to this creative. Need to check whether that
+                     response contains a property that matches the
+                     vendorAnalyticsResponseKey (ex: "priority") for this
+                     custom variable. If so, return the value in the
+                     response object that is associated with that key.
+                  */
+                  const relevant3pResponses = all3pResponses[vendor];
+                  if (relevant3pResponses) {
+                    return relevant3pResponses[
+                        /** @type {string} */
+                        (customVar.vendorAnalyticsResponseKey)];
```

Comment:
I think I might miss something here, but I still don't see how we get the responses for one creative. For example: we have vendor ABC send two responses to two different creatives with response key 'priority'
They are
```
win[AMP_ANALYTICS_3P_RESPONSES] = {
  'ABC': {<'priority': 1>, <'priority' :2>}
}
```
How can I tell which result is for which creative?

Reply:
(Investigating/confirming...will respond later today)

Reply:
Good catch. I've confirmed that we do have a problem here. In the other PR (#11061), only the amp-analytics tag that creates the xframe can call listenFor() on it. See: https://github.com/ampproject/amphtml/blob/ea4b3f60b0c6293d4fb31f100886b666cb1f398b/src/iframe-helper.js#L252

So, the old approach was: have that single recipient (per vendor) store somewhere that all creatives can see, and key off of <vendor, creative>.

The next idea was to just key off of <vendor>, but store the data in a creative-specific way. 

Unfortunately the two got a bit blurred together, and what is in the latest check-in is not correct. What we have now is storing globally but only keying off of vendor. To do the creative-specific storage, we'd have to have the creative who built the xframe pass the message to the actual destination creative window, which sounds overly complex (and any potential solution would involve storing a mapping between creatives on the global window). So, I favor just returning to the previous scenario where something identifying the creative is a component of the response map key.

Reply:
Fixed

Reply:
To confirm we need to store the map in the global window because `<amp-ad-exit>` component lives in the FIE window, while the `<amp-analytics>` component doesn't always live in the FIE window. Could you provide an example that `<amp-analytics>` lives in the ampdoc window, while `<amp-ad-exit>` lives in the FIE window. 

Also regarding the creativeId here, it brings back Keith's [concern](https://github.com/ampproject/amphtml/pull/10714#discussion_r136139913) on the sentinel not exist in friendly iframe.

Reply:
Sure, I can create an example of amp-analytics in ampdoc + amp-ad-exit in FIE. I'm not 100% sure what expected behaviour should be here, since the amp-analytics would presumably not be responding to any creative. Would you say that in this case the amp-ad-exit in FIE should not consider the response to be relevant/actionable?

Regarding the sentinel question...this is tricky.
* Not storing in FIE window because of bug #2942
* If using ampdoc window, need a way to identify the creative, but may not be able to rely on sentinel
At this point, the solutions that come to mind are:
1. Force some kind of unique attribute on the iframe, to use in the way we are currently using the sentinel, which is a bit ugly
2. As you mentioned in the meeting, have the receiving FIE do a second postMessage, which the actual intended FIE can listen for. As Keith mentioned, this is another async. But it would work.
3. Try to find a way around bug #2942 but that sounds like it may be futile.

Thoughts?

Reply:
> Sure, I can create an example of amp-analytics in ampdoc + amp-ad-exit in FIE. I'm not 100% sure what expected behaviour should be here, since the amp-analytics would presumably not be responding to any creative. Would you say that in this case the amp-ad-exit in FIE should not consider the response to be relevant/actionable?

This is exactly my question. If something like this (please ignore the format 😄 ). where `<amp-analytics>` `<amp-ad-exit>` both lives in FIE, there's no question.  And I can totally understand this usecase. We can skip creativeId here because there's only one creative in the FIE window. 
```
<amp-ad id='testad' >
   <FIE>
        <amp-img>Ad Content</amp-img>
        <amp-ad-exit>listen to response from vendor my_type</amp-ad-exit>
        <amp-analytics type='myVendor'></amp-analytics>
   </FIE>
</amp-ad>
```

But I don't know if this is also something we want to support. And what's a valid use case here.
```
<amp-analytics type='myVendor' selector='#testad'>
</amp-analytics>
<amp-ad id='testad' >
   <FIE>
        <amp-img>Ad Content</amp-img>
        <amp-ad-exit>listen to response from vendor my_type</amp-ad-exit>
   </FIE>
</amp-ad>
```

Reply:
To take a step back from amp-ad-exit for a moment: we have spoken to a vendor who is interested in doing 3p amp-analytics on the pub page. But, we (A4A) are considering that out-of-scope for now. And, that particular vendor doesn't have concrete plans to send responses. So, I don't think that use case is necessarily urgent. So my preference would be to have your second example result in the creative ignoring the amp-analytics response from the pub-level amp-analytics tag, at least for now. But if there is demand in the future I am more than happy to do that as a separate project. Does that sound okay?

Reply:
Regarding the sentinel question. 
1) another reason(main reason I think) why we don't have sentinel is that: FIE doesn't need a sentinel to talk to host window. 
2) What about appending the resource id of the parent `<amp-ad>` to the baseURL . Each amp element will have their unique resource id. But the we need some way to pass the `<amp-ad>` resourceId to both `<amp-analytics>` and `<amp-ad-exit>` element.

Reply:
Sounds good to me. Then this may becomes easy. Since there will be only one creative in each FIE, we don't even need a creativeId. right? Can we try to store the response map in the FIE window instead of global ampdoc window?

However if we need to support the second example, we probably will need to come back to this issue again : (

Reply:
I think this resource id would work. I'm not familiar with it, though - do you know where it is defined?
We could do something like this (pseudocode only):
1. In amp-analytics constructor:
```
if (FIE_window['amp-analytics-resourceids']) {
  FIE_window['amp-analytics-resourceids'] = [];
}
push(FIE_window['amp-analytics-resourceids'], this.resourceid);
```
2. When a message is received, store it in ampwin using [vendor][resourceid] as the key
3. When amp-ad-exit determines a click has occurred: look in FIE_window for resourceids of the amp-analytics tags in this FIE. Then for each of those, look in ampwin using [vendor][that resourceid]

Reply:
"Since there will be only one creative in each FIE, we don't even need a creativeId. right?"
There's still the issue that only the FIE which first creates the cross-domain iframe is receiving the messages from that iframe, even though the message may be intended for a different FIE whose creative uses the same vendor.
But I think the resourceid idea will get us around that :)
However, I'm searching through BaseElement and I don't see any field called resourceid?

Reply:
Ah! I knew I missed something here! Now I finally got it
```
<amp-ad id='testad1' >
   <FIE>
        <amp-img>Ad Content</amp-img>
        <amp-ad-exit>listen to response from vendor 'myVendor'</amp-ad-exit>
        <amp-analytics type='myVendor'></amp-analytics>
   </FIE>
</amp-ad>
<amp-ad id='testad2' >
   <FIE>
        <amp-img>Ad Content</amp-img>
        <amp-ad-exit>listen to response from vendor 'myVendor'</amp-ad-exit>
        <amp-analytics type='myVendor'></amp-analytics>
   </FIE>
</amp-ad>
```
Here only FIE of `<amp-ad id='testad1'>` will create the cross domain vendor iframe. You're right we do need creativeId!

The [getResourceId](https://github.com/ampproject/amphtml/blob/master/src/custom-element.js#L1210) method exists in customElement class, but we can add it to BaseElement.

 

> When amp-ad-exit determines a click has occurred: look in FIE_window for resourceids of the amp-analytics tags in this FIE. Then for each of those, look in ampwin using [vendor][that resourceid]

Instead of using the resouceid of the amp-analytics element, I think we should use the resourceid of the parent amp-ad element. Since there can be multiple amp-analytics elements in FIE. 
Also the resourceId is simple a counter that we maintain. In this case, cross-domain iframe can easy send spam response with fake resourceId. Should we worry about this? maybe `${baseurl}-${resourceId}` is a better solution here?

Reply:
Found the resource ID, and created a CL that uses it.